src/pybullet_industrial/g_code_processor.py

```diff
@@ -0,0 +1,435 @@
+import pybullet as p
+from pybullet_industrial import RobotBase
+from pybullet_industrial import linear_interpolation
+from pybullet_industrial import circular_interpolation
+from pybullet_industrial import ToolPath
+import numpy as np
+
+
+class GCodeProcessor:
+    """Initializes a GCodeProcessor object with the provided parameters.
+
+    Args:
+        gcode_input (str, optional): Simulated input G-code
+        robot (RobotBase, optional): The robot which is controlled
+        endeffector_list (list, optional): List of endeffectors to use
+        m_commands (list, optional): M-commands to execute
+        t_commands (list, optional): T-commands to execute
+        offset (np.array, optional): Point which defines the origin
+        axis (int, optional): The axis around which the circle
+                                is interpolated .Defaults to 2 which
+                                corresponds to the z-axis (0=x,1=y)
+        interpolation_precision (int, optional): Precision of interpolation
+        interpolation_approach (int, optional): Number of interpolations
+                                                before determining precision
+    """
+
+    def __init__(self, gcode_input: str = None, robot: RobotBase = None,
+                 endeffector_list: list = None,
+                 m_commands: dict = None,
+                 t_commands: dict = None,
+                 offset: np.array = np.array([[0.0, 0.0, 0.0],
+                                              [0.0, 0.0, 0.0]]),
+                 axis: int = 2, interpolation_precision: int = 0.01,
+                 interpolation_approach: int = 1000):
+
+        #  Converting G-Code into special list format
+        if gcode_input is not None:
+            self.gcode = self.read_gcode(gcode_input)
+
+        # Initializing class variables
+        self.new_point = []
+        self.new_or = []
+        self.last_point = []
+        self.last_or = []
+        self.r_val = 0
+        self.active_endeffector = -1
+        self.robot = robot
+        self.endeffector_list = endeffector_list
+        self.offset = offset
+        self.axis = axis
+        self.interpolation_precision = interpolation_precision
+        self.interpolation_approach = interpolation_approach
+        self.m_commands = m_commands
+        self.t_commands = t_commands
+
+        # Setting the default G-commands
+        self.g_commands = {
+            "54": [lambda: self.g_54()],
+            "500": [lambda: self.g_500()],
+            "17": [lambda: self.g_17()],
+            "18": [lambda: self.g_18()],
+            "19": [lambda: self.g_19()]
+        }
+
+        if robot is not None:
+            self.calibrate_tool()
+
+        if endeffector_list is not None:
```

Comment:
carefull if endeffector_list is None, self.active_endeffector is not set.

This could lead to problems and is also not covered in the tests

Reply:
This change was adapted